proposals/nnnn-span-access-shared-contiguous-storage.md

```diff
@@ -0,0 +1,821 @@
+
+```swift
+extension Span where Element: ~Copyable & ~Escapable {
+  // Unchecked subscripting and extraction
+
+  /// Accesses the element at the specified `position`.
+  ///
+  /// This subscript does not validate `position`; this is an unsafe operation.
+  ///
+  /// - Parameter position: The offset of the element to access. `position`
+  ///     must be greater or equal to zero, and less than `count`.
+  public subscript(unchecked position: Int) -> Element { _read }
+}
+```
+
+##### Index validation utilities:
+
+Every time `Span` uses a position parameter, it checks for its validity, unless the parameter is marked with the word "unchecked". The validation is performed with these functions:
+
+```swift
+extension Span where Element: ~Copyable & ~Escapable {
+  /// Return true if `index` is a valid offset into this `Span`
+  ///
+  /// - Parameters:
+  ///   - index: an index to validate
+  /// - Returns: true if `index` is a valid index
+  public func boundsContain(_ index: Int) -> Bool
+
+  /// Traps if `index` is not a valid offset into this `Span`
+  ///
+  /// - Parameters:
+  ///   - index: an index to validate
+  public func boundsPrecondition(_ index: Int)
+
+  /// Return true if `indices` is a valid range of offsets into this `Span`
+  ///
+  /// - Parameters:
+  ///   - indices: a range of indices to validate
+  /// - Returns: true if `indices` is a valid range of indices
+  public func boundsContain(_ indices: Range<Int>) -> Bool
+
+  /// Traps if `indices` is not a valid range of offsets into this `Span`
+  ///
+  /// - Parameters:
+  ///   - indices: a range of indices to validate
+  public func boundsPrecondition(_ indices: Range<Int>)
+}
+```
+
+Note: these function names are not ideal.
+
+##### Identifying whether a `Span` is a subrange of another:
+
+When working with multiple `Span` instances, it is often desirable to know whether one is a subrange of another. We include a function to determine whether this is the case, as well as a function to obtain the valid offsets of the subrange within the larger span:
+
+```swift
+extension Span where Element: ~Copyable & ~Escapable {
+  /// Returns true if the memory represented by `span` is a subrange of
+  /// the memory represented by `self`
+  ///
+  /// Parameters:
+  /// - span: a span of the same type as `self`
+  /// Returns: whether `span` is a subrange of `self`
+  public func contains(_ span: borrowing Self) -> Bool
+  
+  /// Returns the offsets where the memory of `span` is located within
+  /// the memory represented by `self`
+  ///
+  /// Note: `span` must be a subrange of `self`
+  ///
+  /// Parameters:
+  /// - span: a subrange of `self`
+  /// Returns: A range of offsets within `self`
+  public func offsets(of span: borrowing Self) -> Range<Int>
```

Comment:
[Notes from in person discussions]

1. `func indices(of other: borrowing Self) -> Range<Int>?` appears to be a workable shape/name for this. We should avoid using the term "offset" on `Span` APIs -- `Span` has an `Index`, and these ranges are index ranges. (But `byteOffsets(of:)` would be the corresponding name on `RawSpan`.)
2. Making the return value optional would allow us to remove `isWithin` entirely, getting rid of its alignment surprise. (It seems intuitively obvious that `indices(of:)` would return nil if the other span happens to be misaligned. Not so with `isWithin`.) It is tempting to do precisely this; one drawback is that it seems very easy to skim past `indices(of:)` when one is looking for an operation that implements a structural containment check.